README.md
# NAME

progsche - cool progressbar

# SYNOPSIS
    progsche help
    
    # list available option
    progsche.sh --help

    # run this script  independently
    progsche.sh

    # customize the text yourself while the animation is running and set when the animation should stop
    progsche.sh --custom  $ 100

    # customize your own goodbye text when it's time for animation to stop, it loads a string data type with a maximum limit of 9 strings
    progsche.sh --custom $ 100 The Quick Brown Fox Jumped Over The Lazy Dogs. 

    # display version number
    progsche.sh --version

    # display about this script 
    progsche.sh --about

# DESCRIPTION
Progsche is a progress bar one of the other options you should try to collect on your computer/smartphone although it's not a program that can monitor tasks from commands like cp, mv, dd, tar and others, kind of progress - coreutils.

just a demo program to show general animation bar but improved for entertainment purpose on your desk. Progsche was created as a whimsical project inspired by the DHCP progress bar.

# INSTALLATION
There is no installation trough Recovery mode, just put or push independently.
Just download the file in the latest-Release and open the file manager and move it to your favorite folder. If you do it on the system do your own custom don't forget to set the permission to executetable for it to work.

# USAGE
After all the above processes are done, open your Terminal or Termux:
Type "su" without the quotes then Enter.
Short way
1. NON-PATH Location, path: /sdcard (!! Match where the script file is located!!)

    Type `cd /sdcard` then Enter.
        
    Type again `./progsche.sh` then Enter.

2. PATH Location.

    Type `progsche.sh` then Enter.

# DEPENDENCIES
./

su

mksh

# FEATURE
There are 3 features.

- Colorfull animations
> Many colors on animation
- Customization is your way
> Create your own how the progress bar should display the output and set the stop time as well. Mandatory :-P
- Say Goodbye Outputs
> Can now modify the final appearance if time has run out. Option

# BUG REPORTING

Plese use github issues: http://github.com/luisadha/progsche/

# AUTHOR

Luis Adha <adharudin14@gmail.com>


CHANGELOG
What's new?
15 Nov 2021 [109]
1. Changes - Recreate the display of options on ' -h' to make it easier to read.
2. Changes script descriptlon.
2. Improve Now can insert residual argument with string without limitation.
3. Added Added 1 new option which is '-H' which means more help.

15 Oct 2021 [108]
1. Big change of text in --about . option
2. Changed the unknown option input error message.
3. Big changes to the text in the --help option to make it easier to read.
4. Added 'gnu --help' and others among the options.
5. Increase the maximum string length increased to 9 from 6.
6. Changes to the comments in the script.
7. Release on github

05, Oct 2021 [107]
1. Added the 4th parameter to be filled as a string and displayed it the maximum length of the string is 6
2. Make the 4th parameter condition if it is not filled the default value is displayed.
3. Minor Bug fix
4. Make Busybox unneeded
5. Changing the text from the default values

23 Sep 2019 [105]
1. Create an error message if the -c option is not entered in the Char/String value
2. Also creates an error message if the -c option forgets to enter an Integer value after the Character value
3. Create an error message if the argument $3 is not a numeric value
4. Make the script run without options.

25 Aug 2019 [104]
1. Changing the -e to -c option in the script
2. Support for busybox, busybox is required in skip.
3. Bug fix

03 Jan 2019 [102]
1. Make the script licensed.
2. Create the --about option.
3. Color the animation

31 Dec 2018 [101]
1. Make option -e or --edit
2. Create hidden options
3. Create version options
4. Initial release but not yet published anywhere


progsche.sh
#!/system/bin/sh
#$0
#location /preferdir/thisscript
#for execute you can use this
# cd /preferdir
# ./thisscript
### PROGSCHE: v1.0.8
# Progsche are other progress bar in bash.
#//Progsche adalah Progres bar yang lain di bash.
# Created by <adharudin14@gmail.com>
# Copyright © 2018-2021 Luis Adha
# Need su to be able to run better.
#//Membutuhkan su agar berjalan dengan baik.
### Licenses #####################################
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#  You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
#####################################
BL='\e[01;90m' > /dev/null 2>&1; # HITAM
R='\e[01;91m' > /dev/null 2>&1;  # MERAH
G='\e[01;92m' > /dev/null 2>&1;  # HIJAU
Y='\e[01;93m' > /dev/null 2>&1;  # KUNING
B='\e[01;94m' > /dev/null 2>&1;  # BIRU
P='\e[01;95m' > /dev/null 2>&1;  # UNGU
C='\e[01;96m' > /dev/null 2>&1;  # CYAN
W='\e[01;97m' > /dev/null 2>&1;  # PUTIH
N='\e[0m'     > /dev/null 2>&1;  # NULL
### COLOR #####################################
L='\e[1;38;5;228m'  > /dev/null 2>&1; #LIME
HB='\e[1;38;5;32m'  > /dev/null 2>&1; #HOLO BLUE 
D='\e[0m'           > /dev/null 2>&1; #Default
### ARRAY #####################################
VERSI_ARRAY=( '1.0.8' 'V1.0.8:PROGSCHE' '108' );
NAME_ARRAY=( "progsche" "Progsche" "PROGSCHE" "$0" `busybox basename "$0"` );
RELEASED="Mon, Dec 31";
UPDATED="Fri, Oct 15";
TIMEU="21:06:34 WIB 2021";
TIMER="21:57:03 WIB 2018";
### VARIABLE #####################################
about="
${NAME_ARRAY[1]} v${VERSI_ARRAY[0]} ($UPDATED, $TIMEU) - Copyright (C) 2018 - 2021 by Luis adha.

Description: Other than the common progress or loading bar in bash, this script can be run independently without options, By the way ;-)
Report bugs to <adharudin14@gmail.com>
 ";
help="
Usage: progsche -c CHAR INTEGER ...STRINGS
The -c option requires inputs a character or string if you want and after that it must inputs a number to timeout the animation running.

...
Display any argument input after -c CHAR INTEGER option if any, Remember after integer all inputs are considered strings however it has a limit (default: 9).

 e.g.: progsche --custom $ 150 The quick brown fox jumped over the lazy dogs.


Other options:
  -a   --about    display about script and exit
  -h   --help     display this help message and exit
  -v   --version  output version information and exit

 ";
title="[Presented by PROGSCHE]";
### FUNCTION
#####################################
tentang() {
  clear
        echo "$about";
 echo ''
exit
}




bantuan() {
  clear
      echo "$help";
 echo ''
exit
}




version() {
 clear
 echo "${VERSI_ARRAY[1]}";
exit
}




kursormoves() {
titik=( '' '.' '..' '...' '....' '.....' '......' );
  ####### L1  o2  a3   d4     i5      n6      g7   #7bytes
  #titik harus sama dengan titik2
  titik2=( ${B}'!Loading' ); #silahkan ubah ini maksimal 6huruf kurang dari itu lebih baik, jika Variable titik2 diubah maka variable titik juga harus diubah ini wajib ntar error kursor bergeraknya
}




function default_program() {


 judul="$1"
 waktuhabis=$2

until false; do
clear

  t=0

  kursormoves;

  jarumjamanimsi=( '=' '/' '-' '\' '|' );

    while [ $t -le $waktuhabis ]; do

        for k in "${titik[@]}"
              do
          #clear;
           
        for j in "${jarumjamanimsi[@]}"
              do
        echo -ne ${Y} "\r${judul}${titik2}${Y}${j}${k}" ${D};
          #sleep 0.2;
      

        echo -ne ${Y} "\r${judul}$k$j"
          sleep 0.2

     if [ $t -eq $2 ]; then
        echo -ne ${G}"$title\n"${D};
          sleep 1.5; # clear;
       exit
     fi
     
  let t++
        done

        done
    done
done
}




program="$1";
if [ -z "$program" ]; then
clear
 default_program "$" 175 # Loading(7)=175 #jika 100=4huruf maka 6huruf=150, berapakah 1huruf dalam 'waktuhabis'ny?
#Jawabanya adalaha 25 (stabil)

exit 1

elif [ "$program" == "-?" ] || [ "$program" == "--?" ]; then
clear; bantuan;

elif [ "$program" == "-h" ] || [ "$program" == "-help" ] || [ "$program" == "--help" ]; then
clear; bantuan;

elif [ "$program" == "-a" ] || [ "$program" == "--about" ]; then
clear; tentang;

elif [ "$program" == "-v" ] || [ "$program" == "--version" ]; then
clear; version;

elif [ "$program" == "-c" ] || [ "$program" == "--custom" ];
then
ERROR=${2?Error: option requires an argument -- c }
ERROR2=${3?Error: this option requires more arguments -- c }

int=`echo "$3" | grep -E ^\-?[0-9]*\.?[0-9]+$`

 if [ "$int" != '' ]; then

 judul="$2"
 waktuhabis=$3

until false; do
clear

  t=0

kursormoves;

  jarumjamanimsi=( '=' '/' '-' '\' '|' );

    while [ $t -le $waktuhabis ]; do

        for k in "${titik[@]}"
              do
          #clear;
           
        for j in "${jarumjamanimsi[@]}"
              do
        echo -ne ${Y} "\r${judul}${titik2}${Y}${j}${k}" ${D};
          #sleep 0.2;
      
        echo -ne ${Y} "\r${judul}$k$j"
          sleep 0.2

     if [ $t -eq $3 ]; then
strings="$4 $5 $6 $7 $8 $9 ${10} ${11} ${12}";

papat=$4
 if [ $papat != $4 ]; then
        echo -ne ${G}"$title\n"${D};
          sleep 1.5; # clear;
       exit 0
 else
echo -ne ${G}"${strings[*]}\n"${D};
          sleep 1.5; # clear;
       exit 0
 fi
     fi
     
  let t++
        done

        done
    done
done;


echo $ERROR;
echo $EROOR2;

 else
echo -e "${NAME_ARRAY[0]}: Error: invalid number '$3' ";
  exit 1
 fi
exit 0
else
 clear
    echo -e "${NAME_ARRAY[2]} [HUMAN ERROR DETECTED]: unknown option '$program' ";
echo 'Use -help for a list of options.'
fi
